Builtin resources

Hazardu opened the pull request.

Hazardu committed:
Inital Commit

Moved the changes from the previous branch where everything wenth south.

Hazardu committed:
Added first four shaders

Hazardu committed:
Added UniqueStringLiteralType.h

Hazardu committed:
commit

Hazardu committed:
IFileSystem retrieves data depending on _IRR_EMBED_BUILTIN_RESOURCES_

Hazardu committed:
Creating buffers from data and using them for shaders

UniqueStringLiteralType returns incorrect value in some cases, uncertain why. Some are correct. This causes an issue when retrieving filepath for disabled option to embed resources.
Second issue - generated CPP file needs to be added to Irrlich & IrrlichServer targets. Whats problematic is it being created at build time

Hazardu committed:
generated source files are added to project

devshgraphicsprogramming commented on include/irr/builtin/materials/lambertian/singletexture/specializedshader_fragment:
```diff
@@ -0,0 +1,13 @@
+#version 430 core
```
instead of adding a `_fragment` suffix, do `.frag` its a universally recognized extension

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -178,101 +182,52 @@ void IAssetManager::addLoadersAndWriters()
-{
-    pixelColor = texture(albedo,uv);
-}
-		)===",asset::ISpecializedShader::ESS_FRAGMENT,"irr/builtin/materials/lambertian/singletexture/specializedshader");
```
ok

devshgraphicsprogramming commented on include/irr/builtin/materials/lambertian/singletexture/specializedshader_vertex:
```diff
@@ -0,0 +1,18 @@
+#version 430 core
```
instead of adding a `_vertex` suffix, do `.vert` suffix its a universally recognized extension

devshgraphicsprogramming commented on include/irr/builtin/materials/lambertian/no_texture/specializedshader_fragment:
```diff
@@ -0,0 +1,10 @@
+#version 430 core
```
`.frag`

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -178,101 +182,52 @@ void IAssetManager::addLoadersAndWriters()
-        {
-            pixelColor = vec4(color, 1.0);
-        }
-		)===", asset::ISpecializedShader::ESS_FRAGMENT, "irr/builtin/materials/lambertian/no_texture/specializedshader");
```
ok

devshgraphicsprogramming commented on include/IFileSystem.h:
```diff
@@ -239,6 +241,37 @@ class IFileSystem : public virtual core::IReferenceCounted
 
+
+	template<typename StringUniqueType>
+	inline core::smart_refctd_ptr<asset::ICPUBuffer> loadBuiltinData()
```
`core::ICPUBuffer` you dont want a circular dependency of `asset` namespace on the `io` namespace

Hazardu replied:
since `core::ICPUBuffer` does not exist, I will replace this with `core::IBuffer`

devshgraphicsprogramming replied:
good

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -178,101 +182,52 @@ void IAssetManager::addLoadersAndWriters()
-        layout(location = 0) out vec3 color;
+		buildInShader(fileSystem->loadBuiltinData<IRR_CORE_UNIQUE_STRING_LITERAL_TYPE("irr/builtin/materials/lambertian/no_texture/specializedshader_vertex")>(),
+			asset::ISpecializedShader::ESS_VERTEX, 
+			"irr/builtin/materials/lambertian/no_texture/specializedshader");
```
see my comment about the `.frag` and `.vert` suffices

devshgraphicsprogramming commented on include/IFileSystem.h:
```diff
@@ -239,6 +241,37 @@ class IFileSystem : public virtual core::IReferenceCounted
+		return nullptr;
+#else
+		//in one case, IRR_CORE_UNIQUE_STRING_LITERAL_TYPE value returns an incorrect string
+		//and specializedshader_fragment becomes -> specializedshader_fragmenttt
```
might be because the `IRR_CORE_UNIQUE_STRING_LITERAL_TYPE` macro only works for up to 127 character strings

Hazardu replied:
Length was 72

devshgraphicsprogramming replied:
still a problem?

Hazardu replied:
not a problem anymore

devshgraphicsprogramming commented on include/IFileSystem.h:
```diff
@@ -239,6 +241,37 @@ class IFileSystem : public virtual core::IReferenceCounted
+#else
+		//in one case, IRR_CORE_UNIQUE_STRING_LITERAL_TYPE value returns an incorrect string
+		//and specializedshader_fragment becomes -> specializedshader_fragmenttt
+		auto path = (std::string(__IRR_ROOT_DIRECTORY__) + "/include/" + std::string(StringUniqueType::value));	
```
what is `__IRR_ROOT_DIRECTORY__` ?

devshgraphicsprogramming replied:
ok I see it now

devshgraphicsprogramming commented on include/IFileSystem.h:
```diff
@@ -239,6 +241,37 @@ class IFileSystem : public virtual core::IReferenceCounted
+			file->drop();
+			return retval;	
+		}
+		throw "File cant be opened";
```
DO NOT THROW, IrrlichtBaW is a no-exceptions library!

Just return `nullptr`

devshgraphicsprogramming commented on include/irr/config/BuildConfigOptions.h.in:
```diff
@@ -3,6 +3,9 @@
 #cmakedefine _IRR_DEBUG
 
+// working root directory
+#cmakedefine __IRR_ROOT_DIRECTORY__ "${CMAKE_SOURCE_DIR}"
```
you want `_IRR_INSTALL_DIRECTORY_` and you should be getting the CMAKE installation directory, I will fix it myself afterwards actually

devshgraphicsprogramming replied:
ok this is actually quite a bit complex, but lets do it like this:

1.
When the Irrlicht targets are not installed (all examples and lib run out of the build-dir), this configured header (whose configured version will be placed in `{binary_dir}/include/irr/config/BuildConfigOptions.h` as it is right now, but unlike how it is now, NOT ADDED to the INSTALL target) 

NEEDS to evaluate to `#define _IRR_INSTALL_DIRECTORY_  "{binary_dir}"`

2. When the IRrlicht targets are installed (all examples run out of somewhere, and the library and includes are installed somewhere in a system-wide directory) this configured header will probably have to be placed somewhere out of the way, but added to the INSTALL target.

and the macro NEEDS to evaluate to `#define _IRR_INSTALL_DIRECTORY_ "${CMAKE_INSTALL_PREFIX}"`

Hazardu replied:
Are the files with shader code copied to the install dir as is now?

devshgraphicsprogramming replied:
you should check what the result (resulting files, directories, etc.) is when actually going through with INSTALLing to `Program Files`

devshgraphicsprogramming replied:
actually scap this idea, we don't need this define here

devshgraphicsprogramming commented on src/irr/builtin/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+  # set up dependency
+  add_custom_target(builtin_resources
+   DEPENDS ${CMAKE_BINARY_DIR}/include/irr/builtin/builtinResources.h ${CMAKE_CURRENT_BINARY_DIR}/builtinResourceData.cpp)
+  add_dependencies(Irrlicht builtin_resources)
```
and IrrlichtServer as well needs these deps

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -356,6 +365,7 @@ target_include_directories(Irrlicht PUBLIC
 	"$<$<CONFIG:RELEASE>:${IRRLICHT_CONF_DIR_RELEASE}>"
 )
 target_include_directories(Irrlicht PUBLIC ${IRR_ROOT_PATH}/src ${IRR_ROOT_PATH}/source/Irrlicht ${IRR_ROOT_PATH}/src/3rdparty) # this should be PRIVATE, but things from /src (or /source) are sometimes included in things in /include and so examples have to put source dirs into theirs Include Path
+target_include_directories(Irrlicht PRIVATE ${CMAKE_BINARY_DIR}/include/irr/builtin)
```
you don't need this line at all

Hazardu replied:
it's needed for `#include "builtinResources.h"`

devshgraphicsprogramming replied:
should have been `#include "irr/builtin/builtinResources.h"`

devshgraphicsprogramming replied:
what does it look like now?

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -426,7 +436,7 @@ target_include_directories(IrrlichtServer PUBLIC
 	"$<$<CONFIG:RELEASE>:${IRRLICHT_CONF_DIR_RELEASE}>"
 )
-target_include_directories(IrrlichtServer PRIVATE ${IRR_ROOT_PATH}/src ${IRR_ROOT_PATH}/source/Irrlicht ${IRR_ROOT_PATH}/src/3rdparty)
+target_include_directories(IrrlichtServer PRIVATE ${IRR_ROOT_PATH}/src ${IRR_ROOT_PATH}/source/Irrlicht ${IRR_ROOT_PATH}/src/3rdparty ${CMAKE_BINARY_DIR}/include/irr/builtin)
```
this change is also not needed

devshgraphicsprogramming replied:
and?

Hazardu committed:
Builtin resources

Hazardu committed:
Moved CMakeLists from source/Irrlicht to src/irr

Generated files exist only within a scope the command that generates them is used. It is not possible to add a file to target's sources unless generated file origins from the same cmakelists.txt

Hazardu committed:
fix args in moved function

devshgraphicsprogramming commented on src/irr/CMakeLists.txt:
```diff
@@ -360,8 +392,7 @@ target_include_directories(Irrlicht PUBLIC
 	"$<$<CONFIG:RELEASE>:${IRRLICHT_CONF_DIR_RELEASE}>"
 )
-target_include_directories(Irrlicht PUBLIC ${IRR_ROOT_PATH}/src ${IRR_ROOT_PATH}/source/Irrlicht ${IRR_ROOT_PATH}/src/3rdparty) # this should be PRIVATE, but things from /src (or /source) are sometimes included in things in /include and so examples have to put source dirs into theirs Include Path
-target_include_directories(Irrlicht PRIVATE ${CMAKE_BINARY_DIR}/include)
```
why is the binary include not private?

devshgraphicsprogramming commented on src/irr/CMakeLists.txt:
```diff
@@ -58,6 +75,17 @@ option(_IRR_COMPILE_WITH_GLI_WRITER_ "Compile with GLI Writer" ON)
+	elseif()
+		set(__IRR_INSTALL_DIR__ ${CMAKE_SOURCE_DIR})
+	endif()
+endif()
```
I need to think about this

devshgraphicsprogramming commented on include/IFileSystem.h:
```diff
@@ -260,7 +260,7 @@ class IFileSystem : public virtual core::IReferenceCounted
 		//in one case, IRR_CORE_UNIQUE_STRING_LITERAL_TYPE value returns an incorrect string
 		//and specializedshader_fragment becomes -> specializedshader_fragmenttt
-		auto path = (std::string(__IRR_ROOT_DIRECTORY__) + "/include/" + std::string(StringUniqueType::value));	
+		auto path = (std::string(__IRR_INSTALL_DIR__) + "/include/" + std::string(StringUniqueType::value));
```
I need to see if this will work

devshgraphicsprogramming replied:
we should solve it differently

Hazardu replied:
I am unsure how to pass the reference of SIrrlichtCreationParameter to IFileSystem

Hazardu committed:
Update CMakeLists.txt

Hazardu committed:
Fixes

Hazardu committed:
Wrapped up

Hazardu committed:
Merge remote-tracking branch 'irrbaw/shader_pipeline' into builtin_resources

Hazardu committed:
Removed outdated code from cmakelists
